### DISCUSSION TITLE
[Proposal]: Range shift operators

### DISCUSSION THREAD

**Discussion by SmartK8** [dotnet/csharplang#4634] [Language Ideas]
It would be nice, if ranges could be shifted by an offset in such manner (where `N` is a positive non-zero `Int32` value):

For the right shift:

`..y >> N` will become `..(y + N)` (expand range to the right)
`^x..y >> N` will become `^x..(y + N)` (expand range to the right)
`x..y >> N` will become `(x + N)..(y + N)` (shift range to the right)
`x..^y >> N` will become `(x + N)..^y` (shrink range from the left)
`x.. >> N` will become `(x + N)..` (shrink range from the left)

Analogically for the left shift:

`..y << N` will become `..(y - N)` (shrink range from the right)
`^x..y << N` will become `^x..(y - N)` (shrink range from the right)
`x..y << N` will become `(x - N)..(y - N)` (shift range to the left)
`x..^y << N` will become `(x - N)..^y` (expand range to the left)
`x.. << N` will become `(x - N)..` (expand range to the left)

Other cases: `no change` (compiler warning?)

Basic implementation:

```
public Range Shift(Range range, Int32 value)
{
    if (value != 0 && !(range.Start.IsFromEnd && range.End.IsFromEnd))
    {
        Index start = range.Start.IsFromEnd ? range.Start : range.Start.Value + value;
        Index end = range.End.IsFromEnd ? range.End : range.End.Value + value;

        range = new Range(start, end);
    }

    return range;
}
```

`>>` corresponds to a positive `value`
`<<` corresponds to a negative `value`

The shift operators fit the idea of shifting the whole range and avoid visually clashing with existing operators (mainly `+` and `-`).
I think operators would prevent everyone implementing their own method thus introducing a vector for a mistake.

This would be very useful for data processing, where you have a fixed format but it can be shifted in the read buffer.
But I'm sure there will be more useful scenarios. Using `(x + N)..(y + N)` doesn't allow you to specify fixed `Ranges` ahead of time.

**Comment by TonyValenti** [comment]
That is a neat idea, but I would suggest making an extension method that does that. The reason is because ranges consist of two indexes, and indexes can be either from the start, or from the end.  I am not sure what:
3..^1 >> 5
Would yield. 
3+5=8￼
But what does:
^1+5 = ?

**Reply by SmartK8** [reply]
You're right. I will expand the original post. I think it should be the same as for the fixed end.
